### PR TITLE
Added download support for HLS

### PR DESCRIPTION
**Pull request by ArjixWasTaken**

### CODE REVIEW

File: app/src/main/java/com/lagradost/cloudstream3/utils/M3u8Helper.kt [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+    private fun absoluteExtensionDetermination(url: String): String? {
+        val split = url.split("/")
+        val gg: String = split[split.size - 1].split("?")[0]
+        return if (gg.contains(".")) {
+            gg.split(".")[1].ifEmpty { null }
+        } else null
+    }
+
+    private fun toBytes16BigBecauseArjixIsDumb(n: Int): ByteArray {
```
Comment (Blatzar) [inline app/src/main/java/com/lagradost/cloudstream3/utils/M3u8Helper.kt:24]: Can be replaced with
```kotlin
fun toBytes16Big(n: Int): ByteArray {
        return ByteArray(16) {
            val fixed = n / 256.0.pow((15 - it))
            (maxOf(0, fixed.toInt()) % 256).toByte()
        }
    }
```
I'm pretty sure (haven't tested ByteArray, did chararray before)

File: app/src/main/java/com/lagradost/cloudstream3/utils/M3u8Helper.kt [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+    data class HlsDownloadData(
+        val bytes: ByteArray,
+        val currentIndex: Int,
+        val totalTs: Int,
+        val errored: Boolean = false
+    )
+
+    public fun hlsYield(qualities: List<M3u8Stream>): Iterator<HlsDownloadData> {
+        val selected = selectBest(qualities)!!
```
Comment (Blatzar) [inline app/src/main/java/com/lagradost/cloudstream3/utils/M3u8Helper.kt:130]: ensure that NPE doesn't crash the app if you haven't

File: app/src/main/java/com/lagradost/cloudstream3/utils/M3u8Helper.kt [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+            var encryptionUri: String? = null
+            var encryptionIv = byteArrayOf()
+            var encryptionData= byteArrayOf()
+
+            val encryptionState = isEncrypted(m3u8Data)
+
+            if (encryptionState) {
+                val match = ENCRYPTION_URL_IV_REGEX.find(m3u8Data)!!.destructured
+                encryptionUri = match.component2()
```
Comment (Blatzar) [inline app/src/main/java/com/lagradost/cloudstream3/utils/M3u8Helper.kt:149]: val (_, encryptionUri) = ENCRYPTION_URL_IV_REGEX.find(m3u8Data)!!.destructured
may work iirc

File: app/src/main/java/com/lagradost/cloudstream3/utils/M3u8Helper.kt [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+                            if (encryptionState) {
+                                tsData = getDecrypter(encryptionData, tsData, encryptionIv)
+                                yield(HlsDownloadData(tsData, c, totalTs))
+                                lastYield = c
+                                break
+                            }
+                            yield(HlsDownloadData(tsData, c, totalTs))
+                            lastYield = c
+                        } catch (e: Exception) {
+                            e.printStackTrace()
```
Comment (Blatzar) [inline app/src/main/java/com/lagradost/cloudstream3/utils/M3u8Helper.kt:191]: logError(e)

File: app/src/main/java/com/lagradost/cloudstream3/utils/M3u8Helper.kt (hunk 2 above)
Comment (Blatzar) [reply app/src/main/java/com/lagradost/cloudstream3/utils/M3u8Helper.kt:130]: and in other !! places

File: app/src/main/java/com/lagradost/cloudstream3/utils/M3u8Helper.kt (hunk 3 above)
Comment (ArjixWasTaken) [reply app/src/main/java/com/lagradost/cloudstream3/utils/M3u8Helper.kt:150]: man, look below, I use the match variable elsewhere as well

File: app/src/main/java/com/lagradost/cloudstream3/utils/M3u8Helper.kt (hunk 1 above)
Comment (ArjixWasTaken) [reply app/src/main/java/com/lagradost/cloudstream3/utils/M3u8Helper.kt:25]: Why do I always get an error from that usage of pow, but you dont?
Please use Math.pow instead.
And also no, whilst it works, it has different outputs from my function, so uhh
I don't really need it

Comment (Blatzar) [reply app/src/main/java/com/lagradost/cloudstream3/utils/M3u8Helper.kt:25]: Only diffrence between the functions are your handling of 256 divisible numbers, which probably is due to a bug.

```kotlin
            if (b > 0) {
                tail.add(b.toChar())
            }
```
fucks it up


256 * 10 as input gives
`[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 10]`
while 256 * 10 + 1 gives
`[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 10, 1]`

and regarding the pow stuff, i use `import kotlin.math.pow`

Comment (Blatzar) [reply app/src/main/java/com/lagradost/cloudstream3/utils/M3u8Helper.kt:25]: nvm needs more investigating i fucked up

Comment (ArjixWasTaken) [reply app/src/main/java/com/lagradost/cloudstream3/utils/M3u8Helper.kt:25]: image

**Pull request merged**